File: _B_pendulum/python_qt/plotObserverData.py
import pyqtgraph as pg
import numpy as np


class plotObserverData:
    ''' 
        This class plots the time histories for the pendulum states and state estimates.
    '''

    def __init__(self):
        # Create figure
        self.win = pg.GraphicsWindow()

        # set up plots: rows and columns
        self.z_plot = self.win.addPlot()
        self.win.nextRow()
        self.theta_plot = self.win.addPlot()
        self.win.nextRow()
        self.z_dot_plot = self.win.addPlot()
        self.win.nextRow()
        self.theta_dot_plot = self.win.addPlot()
        # add title and labels
        self.win.setWindowTitle('Pendulum Observer Data')
        self.z_plot.setLabel('left', "position", units='m')
        self.theta_plot.setLabel('left', "angle", units='deg')
        self.z_dot_plot.setLabel('left', "velocity", units='m/s')
        self.theta_dot_plot.setLabel('left', "angular velocity", units='deg/s')
        self.theta_dot_plot.setLabel('bottom', "time", units='s')
        # Use automatic downsampling and clipping to reduce the drawing load
        self.z_plot.setDownsampling(mode='peak')
        self.theta_plot.setDownsampling(mode='peak')
        self.z_dot_plot.setDownsampling(mode='peak')
        self.theta_dot_plot.setDownsampling(mode='peak')
        self.z_plot.setClipToView(True)
        self.theta_plot.setClipToView(True)
        self.z_dot_plot.setClipToView(True)
        self.theta_dot_plot.setClipToView(True)
        # initialize curves and define their colors
        self.z_curve = self.z_plot.plot(pen='b')
        self.zhat_curve = self.z_plot.plot(pen='r')
        self.theta_curve = self.theta_plot.plot(pen='b')
        self.thetahat_curve = self.theta_plot.plot(pen='r')
        self.z_dot_curve = self.z_dot_plot.plot(pen='b')
        self.zhat_dot_curve = self.z_dot_plot.plot(pen='r')
        self.theta_dot_curve = self.theta_dot_plot.plot(pen='b')
        self.thetahat_dot_curve = self.theta_dot_plot.plot(pen='r')

        # Instantiate arrays to hold the time and data histories
        self.time_history = np.empty(100)  # time
        self.z_history = np.empty(100)  # position z
        self.zhat_history = np.empty(100)  # estimated position
        self.theta_history = np.empty(100)  # angle
        self.thetahat_history = np.empty(100)  # estimated angle
        self.z_dot_history = np.empty(100)  # velocity
        self.zhat_dot_history = np.empty(100)  # estimated velocity
        self.theta_dot_history = np.empty(100)  # angular velocity
        self.thetahat_dot_history = np.empty(100)  # estimated angular velocity
        self.ptr = 0  # pointer to current location in histories


    def updatePlots(self, t, x, xhat):
        '''
            Add to the time and data histories, and update the plots.
        '''

        # update the time curve of all plot variables
        self.time_history[self.ptr] = t  # time
        self.z_history[self.ptr] = x[0]  # position
        self.zhat_history[self.ptr] = xhat[0]  # estimated position
        self.theta_history[self.ptr] = 180.0/np.pi*x[1]  # rod angle
        self.thetahat_history[self.ptr] = 180.0/np.pi*xhat[1]  # estimted rod angle
        self.z_dot_history[self.ptr] = x[2]  # velocity
        self.zhat_dot_history[self.ptr] = xhat[2]  # estimated velocity
        self.theta_dot_history[self.ptr] = 180.0/np.pi*x[3]  # angular velocity
        self.thetahat_dot_history[self.ptr] = 180.0/np.pi*xhat[3]  # estimted angular velocity
        self.ptr += 1
        # add data to the buffers if needed, doubling each time
        data_length = self.time_history.shape[0]
        if self.ptr >= data_length:
            tmp = self.time_history
            self.time_history = np.empty(data_length * 2)
            self.time_history[:data_length] = tmp
            tmp = self.z_history
            self.z_history = np.empty(data_length * 2)
            self.z_history[:data_length] = tmp
            tmp = self.zhat_history
            self.zhat_history = np.empty(data_length * 2)
            self.zhat_history[:data_length] = tmp
            tmp = self.theta_history
            self.theta_history = np.empty(data_length * 2)
            self.theta_history[:data_length] = tmp
            tmp = self.thetahat_history
            self.thetahat_history = np.empty(data_length * 2)
            self.thetahat_history[:data_length] = tmp
            tmp = self.z_dot_history
            self.z_dot_history = np.empty(data_length * 2)
            self.z_dot_history[:data_length] = tmp
            tmp = self.zhat_dot_history
            self.zhat_dot_history = np.empty(data_length * 2)
            self.zhat_dot_history[:data_length] = tmp
            tmp = self.theta_dot_history
            self.theta_dot_history = np.empty(data_length * 2)
            self.theta_dot_history[:data_length] = tmp
            tmp = self.thetahat_dot_history
            self.thetahat_dot_history = np.empty(data_length * 2)
            self.thetahat_dot_history[:data_length] = tmp

        # update the plots with associated histories
        self.z_curve.setData(self.time_history[:self.ptr], self.z_history[:self.ptr])
        self.zhat_curve.setData(self.time_history[:self.ptr], self.zhat_history[:self.ptr])
        self.theta_curve.setData(self.time_history[:self.ptr], self.theta_history[:self.ptr])
        self.thetahat_curve.setData(self.time_history[:self.ptr], self.thetahat_history[:self.ptr])
        self.z_dot_curve.setData(self.time_history[:self.ptr], self.z_dot_history[:self.ptr])
        self.zhat_dot_curve.setData(self.time_history[:self.ptr], self.zhat_dot_history[:self.ptr])
        self.theta_dot_curve.setData(self.time_history[:self.ptr], self.theta_dot_history[:self.ptr])
        self.thetahat_dot_curve.setData(self.time_history[:self.ptr], self.thetahat_dot_history[:self.ptr])

File: _B_pendulum/matlab/plotObserverData.m
classdef plotObserverData < handle
    %    This class plots the time histories for the pendulum data.
    %----------------------------
    properties
        % data histories
        time_history
        z_history
        z_hat_history
        theta_history
        theta_hat_history
        z_dot_history
        z_hat_dot_history
        theta_dot_history
        theta_hat_dot_history
        index
        % figure handles
        z_handle
        z_hat_handle
        theta_handle
        theta_hat_handle
        z_dot_handle
        z_hat_dot_handle
        theta_dot_handle
        theta_hat_dot_handle
    end
    methods
        %--constructor--------------------------
        function self = plotObserverData(P)
            % Instantiate lists to hold the time and data histories
            self.time_history = NaN*ones(1,(P.t_end-P.t_start)/P.t_plot);
            self.z_history = NaN*ones(1,(P.t_end-P.t_start)/P.t_plot);
            self.z_hat_history = NaN*ones(1,(P.t_end-P.t_start)/P.t_plot);
            self.theta_history = NaN*ones(1,(P.t_end-P.t_start)/P.t_plot);
            self.theta_hat_history = NaN*ones(1,(P.t_end-P.t_start)/P.t_plot);
            self.z_dot_history = NaN*ones(1,(P.t_end-P.t_start)/P.t_plot);
            self.z_hat_dot_history = NaN*ones(1,(P.t_end-P.t_start)/P.t_plot);
            self.theta_dot_history = NaN*ones(1,(P.t_end-P.t_start)/P.t_plot);
            self.theta_hat_dot_history = NaN*ones(1,(P.t_end-P.t_start)/P.t_plot);
            self.index = 1;

            % Create figure and axes handles
            figure(3), clf
            subplot(3, 1, 1)
                hold on
                self.z_handle = plot(self.time_history, self.z_history, 'b');
                self.z_hat_handle = plot(self.time_history, self.z_hat_history, 'g');
                ylabel('z(m)')
                title('Pendulum State and Observed State')
            subplot(4, 1, 2)
                hold on
                self.theta_handle = plot(self.time_history, self.theta_history, 'b');
                self.theta_hat_handle = plot(self.time_history, self.theta_hat_history, 'g');
                ylabel('theta(deg)')
            subplot(4, 1, 3)
                hold on
                self.z_dot_handle = plot(self.time_history, self.z_dot_history, 'b');
                self.z_hat_dot_handle = plot(self.time_history, self.z_hat_dot_history, 'g');
                ylabel('zdot(m/s)')
            subplot(4, 1, 4)
                hold on
                self.theta_dot_handle = plot(self.time_history, self.theta_dot_history, 'b');
                self.theta_hat_dot_handle = plot(self.time_history, self.theta_hat_dot_history, 'g');
                ylabel('thetadot(deg/s)')
        end
        %----------------------------
        function self = updatePlots(self, t, x, xhat)
            % update the time history of all plot variables
            self.time_history(self.index) = t;
            self.z_history(self.index) = x(1);
            self.theta_history(self.index) = 180/pi*x(2);
            self.z_dot_history(self.index) = x(3);
            self.theta_dot_history(self.index) = 180/pi*x(4);
            self.z_hat_history(self.index) = xhat(1);
            self.theta_hat_history(self.index) = 180/pi*xhat(2);
            self.z_hat_dot_history(self.index) = xhat(3);
            self.theta_hat_dot_history(self.index) = 180/pi*xhat(4);
            self.index = self.index + 1;

            % update the plots with associated histories
            set(self.z_handle, 'Xdata', self.time_history, 'Ydata', self.z_history)
            set(self.theta_handle, 'Xdata', self.time_history, 'Ydata', self.theta_history)
            set(self.z_dot_handle, 'Xdata', self.time_history, 'Ydata', self.z_dot_history)
            set(self.theta_dot_handle, 'Xdata', self.time_history, 'Ydata', self.theta_dot_history)
            set(self.z_hat_handle, 'Xdata', self.time_history, 'Ydata', self.z_hat_history)
            set(self.theta_hat_handle, 'Xdata', self.time_history, 'Ydata', self.theta_hat_history)
            set(self.z_hat_dot_handle, 'Xdata', self.time_history, 'Ydata', self.z_hat_dot_history)
            set(self.theta_hat_dot_handle, 'Xdata', self.time_history, 'Ydata', self.theta_hat_dot_history)
        end
    end
end
